api/app/src/main/kotlin/packit/service/RoleService.kt

```diff
@@ -0,0 +1,98 @@
+package packit.service
+
+import org.springframework.http.HttpStatus
+import org.springframework.security.core.GrantedAuthority
+import org.springframework.security.core.authority.SimpleGrantedAuthority
+import org.springframework.stereotype.Service
+import packit.exceptions.PackitException
+import packit.model.Role
+import packit.model.RolePermission
+import packit.repository.RoleRepository
+
+interface RoleService
+{
+    fun getUsernameRole(username: String): Role
+    fun getAdminRole(): Role
+    fun saveRole(roleName: String)
+    fun checkMatchingRoles(rolesToCheck: List<String>): List<Role>
+    fun getGrantedAuthorities(roles: List<Role>): MutableList<GrantedAuthority>
+}
+
+@Service
+class BaseRoleService(
+    private val roleRepository: RoleRepository
+) : RoleService
+{
+    override fun getUsernameRole(username: String): Role
+    {
+        val userRole = roleRepository.findByName(username)
+
+        if (userRole != null)
+        {
+            return userRole
+        }
+        return roleRepository.save(Role(name = username))
+    }
+
+    override fun getAdminRole(): Role
+    {
+        val userRole = roleRepository.findByName("ADMIN")
+
+        if (userRole != null)
+        {
+            return userRole
+        }
+        return roleRepository.save(Role(name = "ADMIN"))
+    }
```

Comment:
I think we should really be creating built in roles as part of the initial migration, checking if they exist every time we get them seems unnecessary. (Actually, looks like you're already doing that! So I think this save shouldn't be necessary.)

Reply:
yeah i am already doing it but just have a safety net just incase someone were to go and delete it from the table

Reply:
If that happened we'd probably be screwed though, as existing user roles would have been blitzed presumably. It just feels more appropriate here to throw an error if the role doens't exist rather than silently creating it. If the role has been inadvertently deleted, we'll almost certainly have other sorting out to do!

Reply:
yup agree